File: ios/Sources/ACCore/Models/Elements/CompoundButton.swift
import Foundation
// MARK: - CompoundButton

/// Icon descriptor that can be decoded from either a plain string or an object
/// like `{"name": "Calendar", "size": "Small"}`.
public struct IconDescriptor: Codable, Equatable {
    public var name: String
    public var size: String?

    public init(name: String, size: String? = nil) {
        self.name = name
        self.size = size
    }

    public init(from decoder: Decoder) throws {
        // Try as a plain string first
        if let container = try? decoder.singleValueContainer(),
           let stringValue = try? container.decode(String.self) {
            self.name = stringValue
            self.size = nil
            return
        }
        // Otherwise decode as object with name/size
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.name = try container.decode(String.self, forKey: .name)
        self.size = try container.decodeIfPresent(String.self, forKey: .size)
    }
}

public struct CompoundButton: Codable, Equatable {
    public let type: String = "CompoundButton"
    public var id: String?
    public var title: String
    public var description: String?
    public var icon: IconDescriptor?
    public var iconPosition: String?
    public var selectAction: CardAction?
    public var badge: String?
    public var style: String?
    public var isVisible: Bool?
    public var separator: Bool?
    public var spacing: Spacing?
    public var height: BlockElementHeight?
    public var requires: [String: String]?

    /// Convenience accessor: the icon name as a string regardless of how it was encoded
    public var iconName: String? {
        return icon?.name
    }

    public init(
        id: String? = nil,
        title: String,
        description: String? = nil,
        icon: IconDescriptor? = nil,
        iconPosition: String? = nil,
        selectAction: CardAction? = nil,
        badge: String? = nil,
        style: String? = nil,
        isVisible: Bool? = nil,
        separator: Bool? = nil,
        spacing: Spacing? = nil,
        height: BlockElementHeight? = nil,
        requires: [String: String]? = nil
    ) {
        self.id = id
        self.title = title
        self.description = description
        self.icon = icon
        self.iconPosition = iconPosition
        self.selectAction = selectAction
        self.badge = badge
        self.style = style
        self.isVisible = isVisible
        self.separator = separator
        self.spacing = spacing
        self.height = height
        self.requires = requires
    }
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/models/CompoundButton.kt
package com.microsoft.adaptivecards.core.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Icon descriptor that can be decoded from either a plain string (`"Calendar"`)
 * or an object (`{"name": "Calendar", "style": "Regular"}`).
 */
@Serializable(with = IconDescriptorSerializer::class)
data class IconDescriptor(
    val name: String,
    val style: String? = null,
    val size: String? = null
) {
    /** Convenience: the icon name string regardless of how it was encoded */
    override fun toString(): String = name
}

object IconDescriptorSerializer : KSerializer<IconDescriptor> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("IconDescriptor")

    override fun deserialize(decoder: Decoder): IconDescriptor {
        val jsonDecoder = decoder as? JsonDecoder
            ?: return IconDescriptor(name = decoder.decodeString())
        return when (val element = jsonDecoder.decodeJsonElement()) {
            is JsonPrimitive -> IconDescriptor(name = element.content)
            else -> {
                val obj = element.jsonObject
                IconDescriptor(
                    name = obj["name"]?.jsonPrimitive?.content ?: "",
                    style = obj["style"]?.jsonPrimitive?.content,
                    size = obj["size"]?.jsonPrimitive?.content
                )
            }
        }
    }

    override fun serialize(encoder: Encoder, value: IconDescriptor) {
        val jsonEncoder = encoder as? JsonEncoder ?: return
        if (value.style == null && value.size == null) {
            jsonEncoder.encodeJsonElement(JsonPrimitive(value.name))
        } else {
            val obj = buildMap<String, JsonElement> {
                put("name", JsonPrimitive(value.name))
                value.style?.let { put("style", JsonPrimitive(it)) }
                value.size?.let { put("size", JsonPrimitive(it)) }
            }
            jsonEncoder.encodeJsonElement(kotlinx.serialization.json.JsonObject(obj))
        }
    }
}

@Serializable
@SerialName("CompoundButton")
data class CompoundButton(
    @Transient override val type: String = "CompoundButton",
    override val id: String? = null,
    override val isVisible: Boolean = true,
    override val separator: Boolean = false,
    override val spacing: Spacing? = null,
    override val height: BlockElementHeight? = null,
    override val requires: Map<String, String>? = null,
    override val fallback: JsonElement? = null,
    val title: String,
    val description: String? = null,
    val icon: IconDescriptor? = null,
    val iconPosition: String? = null,
    val selectAction: CardAction? = null,
    val badge: String? = null,
    val style: String? = null
) : CardElement {
    /** Convenience accessor: the icon name as a string regardless of how it was encoded */
    val iconName: String? get() = icon?.name
}
